File: src/sven_ros/src/vive_pose_converter_node.cpp
#include <sstream>

#include <vector>

#include "sven_ros/vive_pose_converter_node.h"
#include "jump_detector/datapoint.h"

VivePoseConverterNode::VivePoseConverterNode()
: 	nh(),
	x_bounded(false),
	y_bounded(false),
	z_bounded(false),
	x_min(0),
	x_max(0),
	y_min(0),
	y_max(0),
	z_min(0),
	z_max(0),
	vive_offset_set(false),
	robot_offset_set(false),
	max_vel_passed(false),
	max_window_length(30),
	filter(3, 0, 20),
	stop_button_pressed(true)
{
	pose_pub = nh.advertise<geometry_msgs::PoseStamped>("/vive_pose_converter/pose", 40);
	vive_sub = nh.subscribe("/vive/pose1", 40, &VivePoseConverterNode::vive_pose_callback, this);
	robot_sub = nh.subscribe("/cartesian_pose", 40, &VivePoseConverterNode::robot_callback, this);
	stop_button_sub = nh.subscribe("/vive/controller_LHR_FF6FDF46/joy", 40, &VivePoseConverterNode::stop_button_callback, this);
	
	vive_offset.position.x = 0;
	vive_offset.position.y = 0;
	vive_offset.position.z = 0;
	
	if (nh.getParam("/vive_pose_converter/x_min", x_min)) {
		ROS_INFO_STREAM("X min set to " << x_min);
		x_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/x_max", x_max)) {
		ROS_INFO_STREAM("X max set to " << x_max);
		x_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/y_min", y_min)) {
		ROS_INFO_STREAM("Y min set to " << y_min);
		y_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/y_max", y_max)) {
		ROS_INFO_STREAM("Y max set to " << y_max);
		y_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/z_min", z_min)) {
		ROS_INFO_STREAM("Z min set to " << z_min);
		z_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/z_max", z_max)) {
		ROS_INFO_STREAM("Z max set to " << z_max);
		z_bounded = true;
	}
	if (nh.getParam("/vive_pose_converter/max_vel", max_vel)) {
		ROS_INFO_STREAM("Max vel set to " << max_vel);
	}
	if (nh.getParam("/vive_pose_converter/scale_factor", scale_factor)) {
		ROS_INFO_STREAM("scale_factor set to " << scale_factor);
	}
}

VivePoseConverterNode::~VivePoseConverterNode() {
	return;
}

void VivePoseConverterNode::run() {
	ros::spin();
}

void VivePoseConverterNode::vive_pose_callback(const geometry_msgs::PoseStamped::ConstPtr& msg) {

	if (stop_button_pressed) {
		return;
	}

	ROS_DEBUG_STREAM("Vive message received");
	
	geometry_msgs::PoseStamped vive_pose;
	vive_pose_received(msg, vive_pose);
	
	if (vive_offset_set) {
		geometry_msgs::PoseStamped msg_out;
		if (calculate_pose(msg_out, vive_pose)) {
		
			if (!max_vel_passed && pose_within_velocity_limits(msg_out)) {
				pose_pub.publish(msg_out);
				last_sent_pose = msg_out;
				ROS_DEBUG_STREAM("Message sent");
			}
			else {
				if (!max_vel_passed) {
					ROS_ERROR_STREAM("Message passed max velocity. Restart the node please.");
				}
				max_vel_passed = true;
			}
			max_vel_passed = false;
		}
	}
	else {
		vive_offset.position.x = -vive_pose.pose.position.x;
		vive_offset.position.z = -vive_pose.pose.position.y;
		vive_offset.position.y = -vive_pose.pose.position.z;
		
		vive_offset.orientation.x = -vive_pose.pose.orientation.x;
		vive_offset.orientation.y = -vive_pose.pose.orientation.y;
		vive_offset.orientation.z = -vive_pose.pose.orientation.z;
		vive_offset.orientation.w = -vive_pose.pose.orientation.w;
		
		vive_offset_set = true;
	}
}

void VivePoseConverterNode::robot_callback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
	ROS_DEBUG_STREAM("Robot message received");
	
	if (!robot_offset_set) {
		robot_offset = msg->pose;
		
		// Set last pose sent to robot pose
		bool tmp = vive_offset_set;
		vive_offset_set = false;
		calculate_pose(last_sent_pose, *msg);
		vive_offset_set = tmp;
		
		robot_offset_set = true;
	}
}

void VivePoseConverterNode::stop_button_callback(const sensor_msgs::Joy::ConstPtr& msg) {
	if (msg->buttons[1]) {
		stop_button_pressed = true;
	}
	else {
		stop_button_pressed = false;
	}
}
	

bool VivePoseConverterNode::pose_within_bounds(const geometry_msgs::PoseStamped& msg) {
	if (x_bounded && (msg.pose.position.x < x_min || msg.pose.position.x > x_max)) {
		ROS_DEBUG_STREAM("Outside x bounds");
		return false;
	}
	if (y_bounded && (msg.pose.position.y < y_min || msg.pose.position.y > y_max)) {
		ROS_DEBUG_STREAM("Outside y bounds");
		return false;
	}
	if (z_bounded && (msg.pose.position.z < z_min || msg.pose.position.z > z_max)) {
		ROS_DEBUG_STREAM("Outside z bounds");
		return false;
	}
	return true;
}

bool VivePoseConverterNode::calculate_pose(geometry_msgs::PoseStamped& msg_out, const geometry_msgs::PoseStamped& msg_in) {
	msg_out.header = msg_in.header;
	msg_out.pose = msg_in.pose;
	
	if (vive_offset_set) {
		msg_out.pose.position.y = msg_in.pose.position.z;
		msg_out.pose.position.z = msg_in.pose.position.y;
	
		msg_out.pose.position.x += vive_offset.position.x;
		msg_out.pose.position.y += vive_offset.position.y;
		msg_out.pose.position.z += vive_offset.position.z;
		
		msg_out.pose.position.x *= scale_factor;
		msg_out.pose.position.y *= -scale_factor;
		msg_out.pose.position.z *= scale_factor;
	}
		
	bool result = vive_offset_set && pose_within_bounds(msg_out);
	
	if (robot_offset_set) {
		msg_out.pose.position.x += robot_offset.position.x;
		msg_out.pose.position.y += robot_offset.position.y;
		msg_out.pose.position.z += robot_offset.position.z;
		
		msg_out.pose.orientation.x = robot_offset.orientation.x;
		msg_out.pose.orientation.y = robot_offset.orientation.y;
		msg_out.pose.orientation.z = robot_offset.orientation.z;
		msg_out.pose.orientation.w = robot_offset.orientation.w;
	}
	
	return result && robot_offset_set;
}

bool VivePoseConverterNode::pose_within_velocity_limits(const geometry_msgs::PoseStamped& msg) {
	double max_error = max_vel * (msg.header.stamp.toSec() - last_sent_pose.header.stamp.toSec());
	max_error = max_vel / 30.0;
	return (abs(msg.pose.position.x - last_sent_pose.pose.position.x) < max_error) && (abs(msg.pose.position.y - last_sent_pose.pose.position.y) < max_error) && (abs(msg.pose.position.z - last_sent_pose.pose.position.z) < max_error);
}

bool VivePoseConverterNode::vive_pose_received(const geometry_msgs::PoseStamped::ConstPtr& msg, geometry_msgs::PoseStamped& vive_pose) {
	vive_poses.push(*msg);
	if (vive_poses.size() > max_window_length + 1) {
		vive_poses.pop();
	}
	
	std::vector<DataPoint> xs;
	std::vector<DataPoint> ys;
	std::vector<DataPoint> zs;
	std::vector<DataPoint> qxs;
	std::vector<DataPoint> qys;
	std::vector<DataPoint> qzs;
	std::vector<DataPoint> qws;
	
	std::queue<geometry_msgs::PoseStamped> q = vive_poses;
	
	while (!q.empty()) {
		geometry_msgs::PoseStamped pose_stamped = q.front();
	
		DataPoint x(pose_stamped.header.stamp.toSec(), pose_stamped.pose.position.x);
		DataPoint y(pose_stamped.header.stamp.toSec(), pose_stamped.pose.position.y);
		DataPoint z(pose_stamped.header.stamp.toSec(), pose_stamped.pose.position.z);
		DataPoint qx(pose_stamped.header.stamp.toSec(), pose_stamped.pose.orientation.x);
		DataPoint qy(pose_stamped.header.stamp.toSec(), pose_stamped.pose.orientation.y);
		DataPoint qz(pose_stamped.header.stamp.toSec(), pose_stamped.pose.orientation.z);
		DataPoint qw(pose_stamped.header.stamp.toSec(), pose_stamped.pose.orientation.w);
		xs.push_back(x);
		ys.push_back(y);
		zs.push_back(z);
		qxs.push_back(qx);
		qys.push_back(qy);
		qzs.push_back(qz);
		qws.push_back(qw);
		
		q.pop();
	}
	
	vive_pose.header = msg->header;
	double x;
	double y;
	double z;
	double qx;
	double qy;
	double qz;
	double qw;
	
	if (filter.filter(xs, x)) {
		vive_pose.pose.position.x = x;
	}
	else {
		vive_pose.pose.position.x = msg->pose.position.x;
	}
	if (filter.filter(ys, y)) {
		vive_pose.pose.position.y = y;
	}
	else {
		vive_pose.pose.position.y = msg->pose.position.y;
	}
	if (filter.filter(zs, z)) {
		vive_pose.pose.position.z = z;
	}
	else {
		vive_pose.pose.position.z = msg->pose.position.z;
	}
	if (filter.filter(qxs, qx)) {
		vive_pose.pose.orientation.x = qx;
	}
	else {
		vive_pose.pose.orientation.x = msg->pose.orientation.x;
	}
	if (filter.filter(qys, qy)) {
		vive_pose.pose.orientation.y = qy;
	}
	else {
		vive_pose.pose.orientation.y = msg->pose.orientation.y;
	}
	if (filter.filter(qzs, qz)) {
		vive_pose.pose.orientation.z = qz;
	}
	else {
		vive_pose.pose.orientation.z = msg->pose.orientation.z;
	}
	if (filter.filter(qws, qw)) {
		vive_pose.pose.orientation.w = qw;
	}
	else {
		vive_pose.pose.orientation.w = msg->pose.orientation.w;
	}
	
	return true;
}

int main(int argc, char **argv)
{
	ros::init(argc, argv, "vive_pose_converter");
	VivePoseConverterNode node;
	node.run();
	return 0;
}

File: src/sven_ros/scripts/vive_pose_converter_node.py
#!/usr/bin/env python3

import rospy
import math
from scipy.spatial.transform import Rotation
from std_msgs.msg import Int32
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import PointStamped
from sensor_msgs.msg import Joy

class VivePoseConverterNode(object):
	def __init__(self, vive_scale_factor=1, xlim=(-0.5,0.5), ylim=(-0.5,0.5), zlim=(-0.5,0.5), initialized=False):
		if not initialized:
			rospy.init_node('vive_pose_converter', anonymous=True)
			
		# Settings
		self.vive_scale_factor = vive_scale_factor
		self.xlim = xlim
		self.ylim = ylim
		self.zlim = zlim
		
		# Publishers
		self.pose_pub = rospy.Publisher('/equilibrium_pose', PoseStamped, queue_size=40)
		self.orientation_pub = rospy.Publisher('/equilibrium_orientation', PointStamped, queue_size=40)
		
		# Subscribers
		self.vive_sub = rospy.Subscriber("/vive/pose1", PoseStamped, self.vive_pose_callback)
		self.vive_button_sub = rospy.Subscriber("/vive/controller_LHR_FF6FDF46/joy", Joy, self.vive_button_callback)
		self.robot_sub = rospy.Subscriber("/cartesian_pose", PoseStamped, self.robot_callback)
		
		# Time
		self.rate = rospy.Rate(100)
		
		# Offsets
		self.vive_offset = None
		self.robot_offset = None
		self.button_pressed = True
		
	def run(self):
		rospy.spin()
			
	def vive_pose_callback(self, msg):
		if self.robot_offset is None or self.button_pressed:
			return
			
		vive_pose = []
		
		# Position
		vive_pose.append(msg.pose.position.z * math.cos(math.pi / 4) - msg.pose.position.x * math.sin(math.pi / 4))
		vive_pose.append(msg.pose.position.x * math.cos(math.pi / 4) + msg.pose.position.z * math.sin(math.pi / 4))
		vive_pose.append(msg.pose.position.y)
		
		# Orientation
		or_quat = []
		or_quat.append(msg.pose.orientation.x)
		or_quat.append(msg.pose.orientation.y)
		or_quat.append(msg.pose.orientation.z)
		or_quat.append(msg.pose.orientation.w)
		rot = Rotation.from_quat(or_quat)
		orientation = rot.as_euler('xyz')
		vive_pose.append(orientation[2])
		vive_pose.append(-orientation[0])
		vive_pose.append(-orientation[1] + math.pi / 4)
		
		if self.vive_offset is None:
			self.vive_offset = []
			for i in vive_pose:
				self.vive_offset.append(-i)
		else:
			pose = []
			for i in range(3):
				pose.append((vive_pose[i] + self.vive_offset[i]) * self.vive_scale_factor + self.robot_offset[i])
			for i in range(3,6):
				pose.append(vive_pose[i] + self.vive_offset[i] + self.robot_offset[i])
			
			if pose[0] < xlim[0] or pose[0] > xlim[-1]
				return
			if pose[1] < ylim[0] or pose[1] > ylim[-1]
				return
			if pose[2] < zlim[0] or pose[2] > zlim[-1]
				
			msg_out = PoseStamped()
			msg_out.header = msg.header
			msg_out.pose.position.x = pose[0]
			msg_out.pose.position.y = pose[1]
			msg_out.pose.position.z = pose[2]
			
			or_eul = Rotation.from_euler('xyz',pose[3:6])
			or_quat = or_eul.as_quat()
			msg_out.pose.orientation.x = or_quat[0]
			msg_out.pose.orientation.y = or_quat[1]
			msg_out.pose.orientation.z = or_quat[2]
			msg_out.pose.orientation.w = or_quat[3]
			self.pose_pub.publish(msg_out)
			
			msg_out2 = PointStamped()
			msg_out2.header = msg_out.header
			msg_out2.point.x = pose[3]
			msg_out2.point.y = pose[4]
			msg_out2.point.z = pose[5]
			self.orientation_pub.publish(msg_out2)
			
			
		
	def vive_button_callback(self, msg):
		if msg.buttons[1] == 1:
			self.button_pressed = True
			self.vive_offset = None
		else:
			self.button_pressed = False
		
	def robot_callback(self, msg):
		if self.robot_offset is None:
			self.robot_offset = []
			
			# Position
			self.robot_offset.append(msg.pose.position.x)
			self.robot_offset.append(msg.pose.position.y)
			self.robot_offset.append(msg.pose.position.z)
			
			# Orientation
			or_quat = []
			or_quat.append(msg.pose.orientation.x)
			or_quat.append(msg.pose.orientation.y)
			or_quat.append(msg.pose.orientation.z)
			or_quat.append(msg.pose.orientation.w)
			rot = Rotation.from_quat(or_quat)
			self.robot_offset.extend(rot.as_euler('xyz'))


if __name__ == '__main__':

	rospy.init_node('vive_pose_converter', anonymous=True)
	
	scale_factor = rospy.get_param('~scale_factor')
	x_min = rospy.get_param('~x_min')
	x_max = rospy.get_param('~x_max')
	y_min = rospy.get_param('~y_min')
	y_max = rospy.get_param('~y_max')
	z_min = rospy.get_param('~z_min')
	z_max = rospy.get_param('~z_max')
	
	try:
		node = VivePoseConverterNode(vive_scale_factor=scale_factor, xlim=(x_min,x_max), ylim=(y_min,y_max), zlim=(z_min,z_max), initialized=True)
		node.run()
	except rospy.ROSInterruptException:
		pass
